winpr/libwinpr/clipboard/posix.c

```diff
@@ -572,28 +583,121 @@ static FILEDESCRIPTORW* convert_local_file_list_to_filedescriptors(wArrayList* f
 	return NULL;
 }
 
+static void* convert_any_uri_list_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
+                                                     UINT32* pSize)
+{
+	FILEDESCRIPTORW* descriptors =
+	    convert_local_file_list_to_filedescriptors(clipboard->localFiles);
+
+	WINPR_ASSERT(pSize);
+
+	*pSize = 0;
+	if (!descriptors)
+		return NULL;
+
+	*pSize = (UINT32)ArrayList_Count(clipboard->localFiles) * sizeof(FILEDESCRIPTORW);
+	clipboard->fileListSequenceNumber = clipboard->sequenceNumber;
+	return descriptors;
+}
+
 static void* convert_uri_list_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
                                                  const void* data, UINT32* pSize)
 {
-	FILEDESCRIPTORW* descriptors = NULL;
+	const UINT32 expected = ClipboardGetFormatId(clipboard, mime_uri_list);
+	if (formatId != expected)
+		return NULL;
+	if (!process_uri_list(clipboard, (const char*)data, *pSize))
+		return NULL;
+	return convert_any_uri_list_to_filedescriptors(clipboard, formatId, pSize);
+}
 
-	if (!clipboard || !data || !pSize)
+static BOOL process_files(wClipboard* clipboard, const char* data, UINT32 pSize, const char* prefix)
+{
+	const size_t prefix_len = strlen(prefix);
+
+	WINPR_ASSERT(clipboard);
+
+	ArrayList_Clear(clipboard->localFiles);
+
+	if (!data || (pSize < prefix_len))
+		return FALSE;
+	if (strncmp(data, prefix, prefix_len) != 0)
+		return FALSE;
+	data += prefix_len;
+	pSize -= prefix_len;
+
+	BOOL rc = FALSE;
+	char* copy = strndup(data, pSize);
+	if (!copy)
+		goto fail;
+
+	char* endptr = NULL;
+	char* tok = strtok_s(copy, "\n", &endptr);
```

Comment:
`strtok_s()` is not thread-safe, so this should use the thread safe variant here (`strtok()`) (especially since the posix clipboard can be used outside of FreeRDP (as it is part of the API)).

Reply:
Seems that `strtok()` is not the thread safe variant one. `strtok_r()` is apparently the thread safe one (https://linux.die.net/man/3/strtok).

Reply:
Looks like `strtok_r()` and `strtok_s()` are both fine. Resolving.